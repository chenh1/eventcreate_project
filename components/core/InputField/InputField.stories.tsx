import type { InputFieldProps } from './InputField';
import React from 'react';
import InputField from './InputField';

type Args = {
  args: InputFieldProps
}

const Template = (props: InputFieldProps) => (
  <div className="pt-6">
    <InputField {...props}/>
  </div>
)

export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const InputFieldHyperRed: Args = {
  args: {
    name: "Test",
    highlightColor: "hyper-red",
    placeholder: "First Name",
  },
};

export const InputFieldUltraPurple: Args = {
  args: {
    name: "Test",
    highlightColor: "ultra-purple",
    placeholder: "First Name",
  },
};

export const InputFieldForceYellow: Args = {
  args: {
    name: "Test",
    highlightColor: "force-yellow",
    placeholder: "First Name",
  },
};

export const InputFieldApexBlue: Args = {
  args: {
    name: "Test",
    highlightColor: "apex-blue",
    placeholder: "First Name",
  },
};

export const InputFieldHyperRedError: Args = {
  args: {
    name: "Test",
    highlightColor: "hyper-red",
    placeholder: "First Name",
    isError: true
  },
};

export const InputFieldHyperRedConfirmed: Args = {
  args: {
    name: "Test",
    highlightColor: "hyper-red",
    placeholder: "First Name",
    isConfirmed: true
  },
};