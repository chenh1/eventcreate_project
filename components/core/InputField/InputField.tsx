"use client";

import { getBorderColorClass } from '../../constants/colors';
import type { CoreColors } from '../../constants/colors';
import React, { useMemo, useState, ChangeEvent } from 'react';
import type { UseFormRegister, FieldValues } from 'react-hook-form';

export type InputFieldProps = {
  id?: string;
  value?: string;
  dataCy?: string;
  "data-testid"?: string;
  placeholder?: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  onBlur?: (event: ChangeEvent<HTMLInputElement>) => void;
  onKeyDown?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  disabled?: boolean;
  iconAlign?: 'left' | 'right';
  isError?: boolean;
  isConfirmed?: boolean;
  isInlineBlock?: boolean;
  name: string;
  highlightColor?: CoreColors;
  type?: "text" | "password" | "number" | "email";
  fullWidth?: boolean;
  className?: string;
}

const InputField = React.forwardRef<HTMLInputElement, InputFieldProps>(({
  dataCy,
  id,
  placeholder,
  disabled,
  iconAlign,
  isError,
  isConfirmed,
  isInlineBlock,
  name,
  highlightColor = "max-white",
  type,
  fullWidth,
  onBlur,
  ...props
}, ref: React.ForwardedRef<HTMLInputElement>) => {
  const [ isFocused, setIsFocused ] = useState<boolean>(false);
  const [ inputVal, setInputVal ] = useState<string>('');

  const computedBorderColorClass = useMemo(() => getBorderColorClass(highlightColor), [highlightColor])
  const computedWidthClass = useMemo(() => (fullWidth ? 'w-full' : ''), [fullWidth]);

  return (
    <div className={`relative ${isInlineBlock ? 'inline-block' : ''}`}>
      <label 
        htmlFor={name} 
        className={`pointer-events-none font-headings duration-300 absolute block text-md font-light -top-5 text-sm text-max-white`}
      >
        {placeholder}
      </label>
      <input 
        data-cy={dataCy}
        disabled={disabled}
        onChange={(e) => setInputVal(e.target.value)} 
        onFocus={() => {
          console.log("FOCIUSED")
          setIsFocused(true)} 
        }
        onBlur={(e) => {
          console.log("BLURRING")
          onBlur && onBlur(e)
          setIsFocused(false)} 
        }
        id={id}
        name={name} 
        type={type}
        className={`max-w-full ${computedWidthClass} text-night-black-100 dark:text-max-white font-headings bg-transparent border-b-2 focus:outline-none ${isFocused || inputVal !== '' ? `${computedBorderColorClass}` : 'border-night-black-100 dark:border-max-white'}`}
        ref={ref}
        {...props}
      />
      {isError && <div className={`absolute pointer-events-none ${iconAlign === 'right' ? 'right-0' : 'left-0'} top-0 -translate-x-full`}>&#10060;</div>}
      {isConfirmed && <div className={`absolute pointer-events-none ${iconAlign === 'right' ? 'right-0' : 'left-0'} top-0 -translate-x-full`}>&#x2705;</div>}
    </div>
  )
});

export default InputField