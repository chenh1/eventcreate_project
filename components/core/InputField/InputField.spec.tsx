import InputField from "./InputField";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

it('renders an input text field', async () => {
  const { asFragment } = render(
    <InputField name="text" placeholder="Birthday" />
  );

  expect(asFragment()).toMatchSnapshot();
});



it('renders an input text field with an error state', async () => {
  const { asFragment } = render(
    <InputField name="text" placeholder="Birthday" isError={true}/>
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders an input text field with a confirm state', async () => {
  const { asFragment } = render(
    <InputField name="text" placeholder="Birthday" isConfirmed={true}/>
  );

  expect(asFragment()).toMatchSnapshot();
});


it('renders an input text field with a confirm AND error state if both are available', async () => {
  const { asFragment } = render(
    <InputField name="text" placeholder="Birthday" isError={true} isConfirmed={true}/>
  );

  expect(asFragment()).toMatchSnapshot();
});
