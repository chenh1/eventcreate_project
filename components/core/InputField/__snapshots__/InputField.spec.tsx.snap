// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an input text field 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <label
      class="pointer-events-none font-headings duration-300 absolute block text-md font-light -top-5 text-sm text-max-white"
      for="text"
    >
      Birthday
    </label>
    <input
      class="max-w-full  text-night-black-100 dark:text-max-white font-headings bg-transparent border-b-2 focus:outline-none border-night-black-100 dark:border-max-white"
      name="text"
    />
  </div>
</DocumentFragment>
`;

exports[`renders an input text field with a confirm AND error state if both are available 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <label
      class="pointer-events-none font-headings duration-300 absolute block text-md font-light -top-5 text-sm text-max-white"
      for="text"
    >
      Birthday
    </label>
    <input
      class="max-w-full  text-night-black-100 dark:text-max-white font-headings bg-transparent border-b-2 focus:outline-none border-night-black-100 dark:border-max-white"
      name="text"
    />
    <div
      class="absolute pointer-events-none left-0 top-0 -translate-x-full"
    >
      ❌
    </div>
    <div
      class="absolute pointer-events-none left-0 top-0 -translate-x-full"
    >
      ✅
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders an input text field with a confirm state 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <label
      class="pointer-events-none font-headings duration-300 absolute block text-md font-light -top-5 text-sm text-max-white"
      for="text"
    >
      Birthday
    </label>
    <input
      class="max-w-full  text-night-black-100 dark:text-max-white font-headings bg-transparent border-b-2 focus:outline-none border-night-black-100 dark:border-max-white"
      name="text"
    />
    <div
      class="absolute pointer-events-none left-0 top-0 -translate-x-full"
    >
      ✅
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders an input text field with an error state 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <label
      class="pointer-events-none font-headings duration-300 absolute block text-md font-light -top-5 text-sm text-max-white"
      for="text"
    >
      Birthday
    </label>
    <input
      class="max-w-full  text-night-black-100 dark:text-max-white font-headings bg-transparent border-b-2 focus:outline-none border-night-black-100 dark:border-max-white"
      name="text"
    />
    <div
      class="absolute pointer-events-none left-0 top-0 -translate-x-full"
    >
      ❌
    </div>
  </div>
</DocumentFragment>
`;
